p3/src/block_engine/block_engine.rs

```diff
@@ -0,0 +1,747 @@
+    ) {
+        relayer_metrics.num_removed_connections += disconnected_pubkeys.len() as u64;
+
+        let mut l_subscriptions = subscriptions.write().unwrap();
+        for disconnected in disconnected_pubkeys {
+            if let Some(sender) = l_subscriptions.remove(&disconnected) {
+                datapoint_info!(
+                    "relayer_removed_subscription",
+                    ("pubkey", disconnected.to_string(), String)
+                );
+                drop(sender);
+            }
+        }
+    }
+
+    fn handle_heartbeat(
+        subscriptions: &PacketSubscriptions,
+        relayer_metrics: &mut BlockEngineMetrics,
+    ) -> Vec<Pubkey> {
+        let failed_pubkey_updates = subscriptions
+            .read()
+            .unwrap()
+            .iter()
+            .filter_map(|(pubkey, sender)| {
+                // For BlockEngine protocol, we send empty batches as heartbeats
+                match sender.try_send(Ok(SubscribePacketsResponse {
+                    header: Some(Header {
+                        ts: Some(Timestamp::from(SystemTime::now())),
+                    }),
+                    batch: None, // Empty batch for heartbeat
+                })) {
+                    Ok(_) => {}
+                    Err(TrySendError::Closed(_)) => return Some(*pubkey),
+                    Err(TrySendError::Full(_)) => {
+                        relayer_metrics.num_try_send_channel_full += 1;
+                        warn!("heartbeat channel is full for: {:?}", pubkey);
+                    }
+                }
+                None
+            })
+            .collect();
+
+        relayer_metrics.num_heartbeats += 1;
+
+        failed_pubkey_updates
+    }
+
+    /// Returns pubkeys of subscribers that failed to send
+    fn forward_packets(
+        maybe_packet_batch: Result<PacketBatch, RecvError>,
+        subscriptions: &PacketSubscriptions,
+        slot_leaders: &HashSet<Pubkey>,
+        relayer_metrics: &mut BlockEngineMetrics,
+        _validator_packet_batch_size: usize,
+        forward_all: bool,
+    ) -> BlockEngineResult<Vec<Pubkey>> {
+        let packet_batch = maybe_packet_batch?;
+        let packets: Vec<_> = packet_batch
+            .iter()
+            .filter_map(packet_to_proto_packet)
+            .collect();
+        let batch = ProtoPacketBatch { packets };
+
+        let l_subscriptions: std::sync::RwLockReadGuard<
+            '_,
+            HashMap<Pubkey, TokioSender<Result<SubscribePacketsResponse, Status>>>,
+        > = subscriptions.read().unwrap();
+
+        let senders = if forward_all {
+            l_subscriptions.iter().collect::<Vec<(
+                &Pubkey,
+                &TokioSender<Result<SubscribePacketsResponse, Status>>,
+            )>>()
+        } else {
+            slot_leaders
+                .iter()
+                .filter_map(|pubkey| l_subscriptions.get(pubkey).map(|sender| (pubkey, sender)))
+                .collect()
+        };
```

Comment:
Let's just always forward all. Remove the forward all argument and all logic related to leader/slot tracking